### ISSUE TITLE
Auto-assign UUIDs for concept names and descriptions

### ISSUE THREAD

**Issue by jamlung-ri** [OpenConceptLab/ocl_issues#1683]
Overall: UUIDs should be auto-generated for a concept's names and descriptions. Without it, concepts fail to import into OpenMRS. They currently use the External_ID field for concept names and descriptions, and OCL currently leaves them blank when creating new concept names and descriptions.

Short term fix: OCL gets a method of auto-generating these UUIDs in the External ID field for concept names and descriptions. The method is not yet determined - it might be a "Generate UUID" button in the Concept form, or it might be a Source-level configuration, similar to the existing Auto-assign ID options that currently exist.

Long term solution: 

1. This is part of OCL's repository configuration workflow where a user can designate what IDs they want generated on all of their concepts 
-OR-
2. OCL incorporates true UUIDs into their UUID field for concept names and descriptions. OpenMRS would need to adjust their OCL Module to look to the UUID field for UUIDs, rather than the External ID field. Note that #1671 is a dependency here.

**Comment by jamlung-ri** [comment]
Short term solution:  Add a couple of new options for Source-level configuration, similar to the existing Auto-assign ID options that currently exist. Legacy data will not be affected by this option.

**Comment by jamlung-ri** [comment]
@snyaggarwal let's move the short term solution forward, which likely includes new properties for sources and a couple of new auto-assign ID options in the TermBrowser's source configuration form. Existing sources will default to "None" for configuration schemes.

Please let us know if there are any concerns!

**Comment by snyaggarwal** [comment]
From dev notes:
- We only need “external_id” to have auto-generated UUID and not id/uuid (existing field)
- Note that, similar to a concept’s ID, this can be overridden by a user who wants to manually add an external ID to the concept instead of using OCL’s auto-assigned ID.

**Comment by snyaggarwal** [comment]
@jamlung-ri this is deployed on QA

**Comment by jamlung-ri** [comment]
Thanks Sunny! This is working well! I am seeing auto-generated UUIDs on concept names and descriptions in a test source.  

One quick change - the order of the fields should be as follows: Concept External IDs, Mapping External IDs, Concept Name External IDs, Concept Description External IDs. Basically we should be showing the Concept+Mapping External IDs together, then showing the Name+Description External IDs together.

![image](https://github.com/OpenConceptLab/ocl_issues/assets/55999682/c79d38b5-0f81-4fd6-8bb9-b6315a863bb1)


The only other thing worth noting is that there is no indication that the External ID will be generated, whereas the Concept ID and External ID has some text saying that this field will get autogenerated. Let me bring this up with the OpenMRS Community on Wednesday to see if this is acceptable for now.
![image](https://github.com/OpenConceptLab/ocl_issues/assets/55999682/7de3990b-834d-42eb-9326-4e429f443b1f)

**Comment by bmamlin** [comment]
Another option to consider...

**tl;dr** use javascript to assign a v4 UUID to empty external ID fields when OpenMRS schema is being used

A very hacky (but quick & dirty) fix for this, since we're planning to grow beyond TB v2 anyway, would be to add a `uuid4omrs` class to name & description external ID text fields and throw some javascript in the concept editing form for TB v2 that implements something like:

```javascript
document.addEventListener("DOMContentLoaded", function(event) { 
  var elems = document.getElementsByClassName("uuid4omrs");
  for (var i=0; i < elems.length; i++) {
    if (elems[i].value == "") elems[i].value = crypto.randomUUID();
  };
});
```

If assigning the `uuid4omrs` class name or running the javascript is constrained to source/collection where the OpenMRS Custom Validation Schema is used, only people using OpenMRS sources will get external IDs assigned and they'd get automatically assigned anytime new concepts are created or old ones are edited.

Hacky, yes, but might be a worthy compromise while energy is focused on v3. 🙂

**Comment by jamlung-ri** [comment]
> The only other thing worth noting is that there is no indication that the External ID will be generated, whereas the Concept ID and External ID has some text saying that this field will get autogenerated. Let me bring this up with the OpenMRS Community on Wednesday to see if this is acceptable for now. ![image](https://private-user-images.githubusercontent.com/55999682/289564878-7de3990b-834d-42eb-9326-4e429f443b1f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIzODgzMTcsIm5iZiI6MTcwMjM4ODAxNywicGF0aCI6Ii81NTk5OTY4Mi8yODk1NjQ4NzgtN2RlMzk5MGItODM0ZC00MmViLTkzMjYtNGU0MjlmNDQzYjFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjEyVDEzMzMzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyMTlhNTk4ZTRjZmI3MjE2YzA4ODliYTZmZTU4YTE4NDRjYWRhMjljMThiOGEzMDNiMGQzOTIwN2Q3YzE2ODMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RI4NbhG9uQ4ligEjWTI8-GVZnLqbhGUFLNaEnNLFLXE)

We decided on the Dev call that the final thing to do here is to add text over the External ID fields for Names and Descriptions e.g. “External ID will be auto-assigned (click here to override with manual entry)”

**Comment by snyaggarwal** [comment]
@jamlung-ri this is deployed on QA

**Comment by jamlung-ri** [comment]
Looks great, and it's working great on QA! The OCL/OpenMRS squad likes it too. The only note is that UUIDs won't be automatically assigned to existing concept names unless you edit the concept in TB, but that's no big deal. Let's deploy this!

**Comment by snyaggarwal** [comment]
@jamlung-ri This is deployed on Prod
